xnmt/input.py

```diff
@@ -159,19 +183,41 @@ class BilingualCorpusParser(CorpusParser, Serializable):
 
   yaml_tag = u"!BilingualCorpusParser"
   def __init__(self, src_reader, trg_reader, max_src_len=None, max_trg_len=None, 
-               max_num_train_sents=None, max_num_dev_sents=None):
+               max_num_train_sents=None, max_num_dev_sents=None, sample_train_sents=None):
+    """
+    :param src_reader: InputReader for source side
+    :param trg_reader: InputReader for target side
+    :param max_src_len: filter pairs longer than this on the source side
+    :param max_src_len: filter pairs longer than this on the target side
+    :param max_num_train_sents: only read the first n training sentences
+    :param max_num_dev_sents: only read the first n dev sentences
+    :param sample_train_sents: sample n sentences without replacement from the training corpus (should probably be used with a prespecified vocab)
+    """
     self.src_reader = src_reader
     self.trg_reader = trg_reader
     self.max_src_len = max_src_len
     self.max_trg_len = max_trg_len
     self.max_num_train_sents = max_num_train_sents
     self.max_num_dev_sents = max_num_dev_sents
+    self.sample_train_sents = sample_train_sents
+    if max_num_train_sents is not None and sample_train_sents is not None: raise RuntimeError("max_num_train_sents and sample_train_sents are mutually exclusive!")
 
   def read_training_corpus(self, training_corpus):
     training_corpus.train_src_data = []
     training_corpus.train_trg_data = []
-    for src_sent, trg_sent in six.moves.zip(self.src_reader.read_sents(training_corpus.train_src, self.max_num_train_sents),
-                                            self.trg_reader.read_sents(training_corpus.train_trg, self.max_num_train_sents)):
+    src_len = self.src_reader.count_sents(training_corpus.train_src)
```

Comment:
This will result in always iterating over the corpus twice, right? I think you should only do this when necessary (e.g. when using `sample_train_sents`). It would be even better if we only had to do it once at the beginning of training (e.g. cache it in something like `training_corpus.train_src_len` and re-use the cached value).